wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java

```diff
@@ -0,0 +1,306 @@
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.Page;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.devfile.server.model.impl.UserDevfileImpl;
+import org.eclipse.che.api.devfile.server.spi.UserDevfileDao;
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.core.db.jpa.DuplicateKeyException;
+import org.eclipse.che.core.db.jpa.IntegrityConstraintViolationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+@Beta
+public class JpaUserDevfileDao implements UserDevfileDao {
+  private static final Logger LOG = LoggerFactory.getLogger(JpaUserDevfileDao.class);
+
+  @Inject private Provider<EntityManager> managerProvider;
+  public static final List<Pair<String, String>> DEFAULT_ORDER =
+      ImmutableList.of(new Pair<>("id", "ASC"));
+
+  @Override
+  public UserDevfileImpl create(UserDevfileImpl devfile) throws ConflictException, ServerException {
+    requireNonNull(devfile);
+    try {
+      doCreate(devfile);
+    } catch (DuplicateKeyException ex) {
+      throw new ConflictException(
+          format("Devfile with name '%s' already exists for current user", devfile.getName()));
+    } catch (IntegrityConstraintViolationException ex) {
+      throw new ConflictException(
+          "Could not create devfile with creator that refers on non-existent user");
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getLocalizedMessage(), ex);
+    }
+    return new UserDevfileImpl(devfile);
+  }
+
+  @Override
+  public UserDevfileImpl update(UserDevfileImpl update)
+      throws NotFoundException, ConflictException, ServerException {
+    requireNonNull(update);
+    try {
+      return new UserDevfileImpl(doUpdate(update));
+    } catch (DuplicateKeyException ex) {
+      throw new ConflictException(
+          format("Devfile with name '%s' already exists for current user", update.getName()));
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getLocalizedMessage(), ex);
+    }
+  }
+
+  @Override
+  public void remove(String id) throws ServerException {
+    requireNonNull(id);
+    try {
+      doRemove(id);
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getLocalizedMessage(), ex);
+    }
+  }
+
+  @Override
+  @Transactional(rollbackOn = {ServerException.class})
+  public UserDevfileImpl getById(String id) throws NotFoundException, ServerException {
+    requireNonNull(id);
+    try {
+      final UserDevfileImpl devfile = managerProvider.get().find(UserDevfileImpl.class, id);
+      if (devfile == null) {
+        throw new NotFoundException(format("Devfile with id '%s' doesn't exist", id));
+      }
+      return new UserDevfileImpl(devfile);
```

Comment:
Isn't the `devfile` detached from the JPA session when it leaves this method anyway? Why do we need to create a copy of it?

Reply:
AFAIK we don't use partial updates of the object. We replace the whole object. By design we not used attached objects at all.